spacy/matcher/phrasematcher.pyx

```diff
@@ -345,6 +342,8 @@ def unpickle_matcher(vocab, docs, callbacks, attr):
     matcher = PhraseMatcher(vocab, attr=attr)
     for key, specs in docs.items():
         callback = callbacks.get(key, None)
+        # TODO is this correct?
+        specs = list(specs)
```

Comment:
I haven't dug down to the root of this, but I assume this is just a quirk of serialization and it's OK to do this. I can look into it more if it's surprising.

Reply:
Fix this starting at `__reduce__` instead to work with `sorted(dict.items())`.

Reply:
(I'm not entirely sure what the error in the tests was, but there aren't any sets here.)

Reply:
As part of this PR I added a check that the list of Docs to `PhraseMatcher.add` is a `List` (`isinstance(docs, List)`), and that fails with an error that the variable is a `set`. Let me look at what's going on more...

Reply:
OK, I figured out the error. 

The PhraseMatcher stores things to match in `self._docs`, which is of type `Dict[str, Set[Tuple[attr_t]]]`. The input `docs` can be a list of Docs, which would be then be converted for internal storage, or the pre-processed values stored in this internal dict. The input Doc objects are not saved anywhere, so serialization is using these sets.

I will loosen the type check for now, though the multi-type stuff here is a little confusing.

Reply:
That is weirder than I remembered.

Reply:
What we could do, if we wanted the types/interface to be more uniform, is have a separate method that just sets these processed values. It could be called something like `_set_internal_doc_representations` or `_add_processed_patterns` and would normally only be used in deserialization. It would just be an assignment. 

I don't think either way is cleaner so my inclination would be just to leave things as they are.

Reply:
It is kind of messy and since we were both confused, I think it probably would be worth it to add a function to do this. Maybe `_add_from_arrays` or something like that? I guess you'd have to move the actual adding to another private method, too, hmm...

Reply:
OK, I've changed this to have an `_add_from_array` function, which takes a key and the pattern list. This directly replaces the last bit of `add`. For unpickling I added `_add_from_arrays`, which is not quite just a for loop over the single function - it's more like `add` without preprocessing/safety features, in that it takes a callback. 

One change that isn't strictly necessary here is that I made `self._docs` a `defaultdict` - since `setdefault` was being called for every key, this shouldn't change the behaviour at all, and it means one less thing to initialize when adding.

While working on this, one thought I had was that it would be nice to replace part of `add` with a call to the batch method, which would remove the need for the single add method. However I realized that doesn't match the behaviour of `add` at the moment - if you have multiple patterns, and one of them is invalid, everything up until the invalid pattern will be added (because the items are verified and then added immediately, one at a time). That seems kind of weird, should we change it to verify all the patterns before adding any of them?

Reply:
I think it would be okay to change the behavior to do all the validation first.

Reply:
OK, I have changed this to do the validation first.